ksql-engine/src/main/java/io/confluent/ksql/query/KafkaStreamsBuilderImpl.java

```diff
@@ -21,23 +21,25 @@
 import org.apache.kafka.streams.KafkaClientSupplier;
 import org.apache.kafka.streams.KafkaStreams;
 import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.Topology;
 
 public class KafkaStreamsBuilderImpl implements KafkaStreamsBuilder {
 
   private final KafkaClientSupplier clientSupplier;
 
-  public KafkaStreamsBuilderImpl(final KafkaClientSupplier clientSupplier) {
-    Objects.requireNonNull(clientSupplier, "clientSupplier can't be null");
-    this.clientSupplier = clientSupplier;
+  KafkaStreamsBuilderImpl(final KafkaClientSupplier clientSupplier) {
+    this.clientSupplier = Objects.requireNonNull(clientSupplier, "clientSupplier");
   }
 
   @Override
-  public KafkaStreams buildKafkaStreams(
+  public BuildResult buildKafkaStreams(
       final StreamsBuilder builder,
-      final Map<String, Object> conf) {
-
+      final Map<String, Object> conf
+  ) {
     final Properties props = new Properties();
     props.putAll(conf);
-    return new KafkaStreams(builder.build(), props, clientSupplier);
+    final Topology topology = builder.build(props);
```

Comment:
Importantly, we now pass stream props to the topology building, so that it can read `topology.optimize` prop